### PR TITLE
Crt release v1

### PR DESCRIPTION
**Pull request by dobertRowneySr**
This PR contains all the fixes for v1 in order for allow Orion (and Atlas) to progress
#### Changes (and issues addressed)
- [x]  remove symbol map from runtime #4789 
- [x] Avoid that Amm and revenue split being active #4783
- [x] update all CRT metadata #4713 
- [x] remove the deadline parameter from amm extr.s #4483 
- [x] both module account and amm treasury account are guaranteed to be initialised with `ExistentialDeposit` #4198 
- [x] update with nara #4876
- [x] benchmark Amm extrinsics #4446 
- [x] merge changes for #4378
- [x] Added minor security fixes such as make the `TransferOutput` parameter a bounded vector and add a minimum allowed value for the Amm `a`
- [x] Also after carefully reviewing AMM parameters default recommendations https://github.com/Joystream/joystream/issues/4488#issuecomment-1768635722 Settled on types for amm curve parameters

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1204404358291458/1205679220762292) by [Unito](https://www.unito.io)

### CODE REVIEW

File: types/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@joystream/types",
-  "version": "2.1.0",
+  "version": "2.3.0",
```
Comment (mnaamani) [inline types/package.json]: I know I already suggested 2.3.0, but seeing that nara will also need a version bump. Lets update major version instead (I think this is what we came to as best approach with Leszek)

```suggestion
  "version": "4.0.0",
```

And Nara will be version `3.0.0`

File: distributor-node/package.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
     "@joystream/metadata-protobuf": "^2.8.0",
     "@joystream/opentelemetry": "1.0.0",
     "@joystream/storage-node-client": "^3.0.0",
-    "@joystream/types": "^3.0.0",
+    "@joystream/types": "^2.0.0",
```
Comment (mnaamani) [inline distributor-node/package.json]: Not clear why this is being reverted .. but I can imagine it was quite difficult to keep things in sync with the multitude of times you rebased

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -240,16 +240,9 @@ impl Contains<<Runtime as frame_system::Config>::RuntimeCall> for CallFilter {
             RuntimeCall::Content(content::Call::<Runtime>::update_channel_privilege_level {
                 ..
             }) => false,
-            RuntimeCall::Content(content::Call::<Runtime>::update_channel_nft_limit { .. }) => {
```
Comment (mnaamani) [inline runtime/src/lib.rs]: Is it essential to enable `update_channel_nft_limit` for creator tokens to work?

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -240,16 +240,9 @@ impl Contains<<Runtime as frame_system::Config>::RuntimeCall> for CallFilter {
...
-                false
-            }
-            RuntimeCall::Content(
```
Comment (mnaamani) [inline runtime/src/lib.rs]: Is `set_channel_paused_features_as_moderator` essential for creator tokens to be released?

File: tests/network-tests/src/scenarios/content-directory.ts [hunk 5]
```diff
@@ -33,5 +33,6 @@ scenario('Content directory', async ({ job }) => {
     'curators and collaborators permissions',
     collaboratorAndCuratorPermissions
   ).after(videoCommentsAndReactionsJob)
-  job('direct channel payment by members', directChannelPayment).after(collaboratorAndCuratorPermissionsJob)
+  // TODO fix: https://github.com/Joystream/joystream/issues/4756
```
Comment (mnaamani) [inline tests/network-tests/src/scenarios/content-directory.ts]: Seems this comment can be removed now?

File: types/package.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@joystream/types",
-  "version": "3.0.0",
-  "description": "Types for Joystream Substrate Runtime - Nara release",
+  "version": "2.2.0",
```
Comment (mnaamani) [inline types/package.json]: Please revert back to 3.0.0. Hmm didn't we publish 4.0.0 which was originally for CRT.. In that case this will need to be 4.1.0 we will just have to skip 3.x

File: query-node/chain-metadata/1001.json
Comment (mnaamani) [inline query-node/chain-metadata/1001.json:1]: What happened here? This file should not change.

File: query-node/chain-metadata/2001.json
Comment (mnaamani) [inline query-node/chain-metadata/2001.json:1]: What happened here? This file should not change.

File: query-node/mappings/src/content/channel.ts [hunk 7]
```diff
@@ -86,14 +87,17 @@ import {
...
 
 export async function content_ChannelCreated(ctx: EventContext & StoreContext): Promise<void> {
-  const { store, event } = ctx
+  const { store, event, block } = ctx
+  const { specVersion } = block.runtimeVersion
+
   // read event data
   const [channelId, { owner, dataObjects, channelStateBloatBond }, channelCreationParameters, rewardAccount] =
-    new ChannelCreatedEvent_V1001(event).params
+    specVersion < 2002 ? new ChannelCreatedEvent_V1001(event).params : new ChannelCreatedEvent_V2002(event).params
```
Comment (mnaamani) [inline query-node/mappings/src/content/channel.ts]: It doesn't look like the `ChannelCreated` event was changed in this PR.. or `nara`, So why is this needed? Is it related possible to `1001.json` being modified? Perhaps since work on crt was started such a long time ago, the spec version was 1001 at that time..?

File: query-node/mappings/src/content/channel.ts [hunk 8]
```diff
@@ -160,9 +164,12 @@ export async function content_ChannelCreated(ctx: EventContext & StoreContext):
...
 
 export async function content_ChannelUpdated(ctx: EventContext & StoreContext): Promise<void> {
-  const { store, event } = ctx
+  const { store, event, block } = ctx
+  const { specVersion } = block.runtimeVersion
+
   // read event data
-  const [, channelId, channelUpdateParameters, newDataObjects] = new ChannelUpdatedEvent_V1001(event).params
+  const [, channelId, channelUpdateParameters, newDataObjects] =
+    specVersion < 2002 ? new ChannelUpdatedEvent_V1001(event).params : new ChannelUpdatedEvent_V2002(event).params
```
Comment (mnaamani) [inline query-node/mappings/src/content/channel.ts]: Same issue here I think, the 1001,2001 .json files are incorrect.

**Review by mnaamani — CHANGES_REQUESTED** [review]
I had to stop in middle of review to bring up mainly issue with changes being done in query-node mapping. They are not wrong, but seems most of the changes related to picking the event type based on runtime spec version is because you have wrong metadata in files 1001.json and 2001.json

Can we confirm this is not a general issue for event types in nara related to new version of substrate?

File: distributor-node/package.json (hunk 2 above)
Comment (dobertRowneySr) [reply distributor-node/package.json]: It was due to the `yarn build:packages` failing

File: runtime/src/lib.rs (hunk 4 above)
Comment (dobertRowneySr) [reply runtime/src/lib.rs]: No, it is probably due to older modifications. 
I have it re-filtered again

File: runtime/src/lib.rs (hunk 3 above)
Comment (dobertRowneySr) [reply runtime/src/lib.rs]: No, it is probably due to older modifications. 
I have it re-filtered again

File: types/package.json (hunk 6 above)
Comment (dobertRowneySr) [reply types/package.json]: I have noticed that setting "^4.0.0" doesn't cause the CI to break
https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: distributor-node/package.json (hunk 2 above)
Comment (dobertRowneySr) [reply distributor-node/package.json]: Setting it to ^4.0.0 fixes the CI

File: query-node/chain-metadata/1001.json
Comment (dobertRowneySr) [reply query-node/chain-metadata/1001.json:1]: fixed by copying 1001.json and 2001.json from nara https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: query-node/chain-metadata/2001.json
Comment (dobertRowneySr) [reply query-node/chain-metadata/2001.json:1]: fixed by copying 1001.json and 2001.json from nara https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: query-node/mappings/src/content/channel.ts (hunk 8 above)
Comment (dobertRowneySr) [reply query-node/mappings/src/content/channel.ts]: fixed by copying 1001.json and 2001.json from nara https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: query-node/mappings/src/content/channel.ts (hunk 7 above)
Comment (dobertRowneySr) [reply query-node/mappings/src/content/channel.ts]: fixed by copying 1001.json and 2001.json from nara https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: tests/network-tests/src/scenarios/content-directory.ts (hunk 5 above)
Comment (dobertRowneySr) [reply tests/network-tests/src/scenarios/content-directory.ts]: fixed in https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: runtime/src/lib.rs (hunk 4 above)
Comment (dobertRowneySr) [reply runtime/src/lib.rs]: here https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: runtime/src/lib.rs (hunk 3 above)
Comment (dobertRowneySr) [reply runtime/src/lib.rs]: here https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: distributor-node/package.json (hunk 2 above)
Comment (dobertRowneySr) [reply distributor-node/package.json]: here https://github.com/Joystream/joystream/pull/4871/commits/ead0209600d711681888c6aff1ed222e07422b4e

File: runtime-modules/content/src/lib.rs [hunk 9]
```diff
@@ -3675,6 +3633,98 @@ decl_module! {
...
+            // Call to ProjectToken
+            T::ProjectToken::deactivate_amm(
+                token_id,
+                member_id,
+            )?;
+
+        }
+
+        /// Allow crt issuer to update metadata for an existing token
+        #[weight = 100_000_000] // TODO: Adjust weight
```
Comment (mnaamani) [inline runtime-modules/content/src/lib.rs]: Do we need a benchmarked weight function for this?

File: query-node/mappings/src/content/channel.ts (hunk 7 above)
Comment (mnaamani) [reply query-node/mappings/src/content/channel.ts]: Ok, but technically the code now is suggesting that there is a difference in types. Its not a big deal but it might confuse the reader.

File: query-node/mappings/src/content/channel.ts (hunk 8 above)
Comment (mnaamani) [reply query-node/mappings/src/content/channel.ts]: Same here, v1001 of the event will work in nara.

File: query-node/mappings/src/proposals.ts [hunk 10]
```diff
@@ -333,9 +337,15 @@ async function handleRuntimeUpgradeProposalExecution(event: SubstrateEvent, stor
...
-export async function proposalsCodex_ProposalCreated({ store, event }: EventContext & StoreContext): Promise<void> {
+export async function proposalsCodex_ProposalCreated({
+  store,
+  event,
+  block,
+}: EventContext & StoreContext): Promise<void> {
+  const specVersion = block.runtimeVersion.specVersion
   const [proposalId, generalProposalParameters, runtimeProposalDetails, proposalThreadId] =
-    new ProposalCreatedEvent_V1001(event).params
+    specVersion >= 2002 ? new ProposalCreatedEvent_V2002(event).params : new ProposalCreatedEvent_V1001(event).params
```
Comment (mnaamani) [inline query-node/mappings/src/proposals.ts:347]: Here again, the `ProposalCreated` event hasn't changed since v1001, but since we are planning a new proposal type to be added for freezing, `ProposalDetails` will change (which is one of the data elements of `ProposalCreated`) we can keep this change.

File: runtime-modules/project-token/src/lib.rs [hunk 11]
```diff
@@ -1622,6 +1848,11 @@ impl<T: Config> Module<T> {
         <T as Config>::ModuleId::get().into_sub_account_truncating(Vec::<u8>::new())
     }
 
+    /// Returns the account for the AMM treasury
+    pub fn amm_treasury_account(token_id: T::TokenId) -> T::AccountId {
```
Comment (mnaamani) [inline runtime-modules/project-token/src/lib.rs:1914]: This is a new "module" account that was not initialized at genesis. I think we need a runtime migration to give it an existential deposit.

File: runtime-modules/project-token/src/lib.rs [hunk 12]
```diff
@@ -161,11 +146,27 @@ decl_storage! { generate_storage_info
...
+        pub AmmBuyTxFees get(fn amm_buy_tx_fees) config(): Permill;
+
+        /// AMM sell transaction fee percentage
+        pub AmmSellTxFees get(fn amm_sell_tx_fees) config(): Permill;
+
+        /// Max patronage rate allowed
+        pub MaxYearlyPatronageRate get(fn max_yearly_patronage_rate) config(): YearlyRate = YearlyRate(Permill::from_percent(15));
+
+        /// Minimum slope parameters allowed for AMM curve
+        pub MinAmmSlopeParameter get(fn min_amm_slope_parameter) config(): TokenBalanceOf<T>;
```
Comment (mnaamani) [inline runtime-modules/project-token/src/lib.rs]: Just to be clear, all these new storage values will have value of zero (with exception of `max_yearly_patronage_rate`) after the runtime upgrade as they are not initialized at genesis.
If we need them to have some other initial value we will want a runtime migration.

**Review by mnaamani — CHANGES_REQUESTED** [review]
Left a few more comments.
The most important issue however is the need to initialize the `amm_treasury_account` with existential deposit which we do for all module accounts, with an on_runtime_update migration.

File: types/package.json (hunk 6 above)
Comment (mnaamani) [reply types/package.json]: So that would be using the published version to npm (the yarn.lock is modified), not the version in the monorepo.
Setting the dependency in cli to ^4.0.0 is okay, but the types package in the monorepo should be updated to 4.1.0 then probably need to revert change in yarn.lock

File: runtime-modules/project-token/src/migrations.rs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+        fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+            let onchain = Pallet::<T>::on_chain_storage_version();
+
+            ensure!(onchain < 1, "this migration can be deleted");
+
+            let module_account_id = Module::<T>::module_treasury_account();
+            let deposit = T::JoyExistentialDeposit::get();
+            let treasury_free_balance = Joy::<T>::free_balance(&module_account_id);
+
+            ensure!(treasury_free_balance >= Joy::<T>::existential_deposit());
```
Comment (mnaamani) [inline runtime-modules/project-token/src/migrations.rs]: Does it make sense to assert this? The whole point is that this is a new account which would have no free balance.. So this will fail when run against mainnet chain..If anything you would be asserting the free balance is in-fact zero.

File: runtime-modules/project-token/src/migrations.rs [hunk 14]
```diff
@@ -0,0 +1,78 @@
+use super::*;
+use frame_support::traits::OnRuntimeUpgrade;
+
+// syntactic sugar for logging.
+#[macro_export]
+macro_rules! log {
+	($level:tt, $patter:expr $(, $values:expr)* $(,)?) => {
+		log::$level!(
```
Comment (mnaamani) [inline runtime-modules/project-token/src/migrations.rs]: I think you are missing the log crate dependency in Cargo.toml and LOG_TARGET definition

File: runtime-modules/project-token/src/migrations.rs (hunk 13 above)
Comment (mnaamani) [reply runtime-modules/project-token/src/migrations.rs]: Besides that, the ensure!() macro needs a second argument which is the error if assertion fails.

File: yarn.lock [hunk 15]
```diff
@@ -2347,6 +2347,19 @@
     yaml "^1.10.0"
     yaml-validator "^3.0.0"
 
+"@joystream/types@^4.0.0":
+  version "4.0.0"
+  resolved "https://registry.yarnpkg.com/@joystream/types/-/types-4.0.0.tgz#b7019466e5fb82fe4a97d46fcbcf840e19ab3ff6"
+  integrity sha512-3NE+sQC9cXBudYIPrCJCDdM4N9p6a4/XLUCjaZLaHwiD1Ds7Dbsfwjbfw67wU1m2V9hUeQCacV/ixumjcOIBsw==
```
Comment (mnaamani) [inline yarn.lock]: run `yarn` in the root to remove this entry in yarn.lock

File: runtime/src/lib.rs [hunk 16]
```diff
@@ -901,6 +900,7 @@ parameter_types! {
     pub const ProjectTokenModuleId: PalletId = PalletId(*b"mo:token"); // module: token
     pub const MaxVestingSchedulesPerAccountPerToken: u32 = 5;
     pub const BlocksPerYear: u32 = 5259600; // 365,25 * 24 * 60 * 60 / 6
+    pub const MaxOutputs: u32 = 256; // TODO(Martin, Ignazio) : find a suitable value
```
Comment (mnaamani) [inline runtime/src/lib.rs]: Can you make an issue for this if this is not the final value?

File: tsconfig.json [hunk 17]
```diff
@@ -1,3 +1,22 @@
 {
-
+    "compilerOptions": {
```
Comment (mnaamani) [inline tsconfig.json]: What problem is this solving?

File: runtime-modules/project-token/Cargo.toml [hunk 18]
```diff
@@ -5,28 +5,35 @@ authors = ['Joystream contributors']
...
+codec = { package = 'parity-scale-codec', version = '3.1.2', default-features = false, features = [
+	'derive',
+] }
+serde = { version = '1.0.101', features = ['derive'], optional = true }
+scale-info = { version = "2.1.1", default-features = false, features = [
+	"derive",
+] }
+
+# util
+log = { version = "0.4.17", default-features = false }
```
Comment (mnaamani) [inline runtime-modules/project-token/Cargo.toml]: This is no longer used, it was used at one point in the migrations.rs module which has now been removed.

**Review by mnaamani — CHANGES_REQUESTED** [review]
LGTM
just some minor points.

**Review by mnaamani — APPROVED** [review]
